### PR TITLE
[ModelZoo] Refactor ERNIE-M usage in Model Zoo

### PR DESCRIPTION
**Pull request by Yam0214**
### PR types

Function optimization

### PR changes

APIs

### Description

Refactor ERNIE-M usage in Model Zoo and add unittest.

### CODE REVIEW

File: model_zoo/ernie-m/README.md [hunk 1]
```diff
@@ -60,22 +61,31 @@ XNLI 是 MNLI 的子集，并且已被翻译成14种不同的语言（包含一
...
     --task_type cross-lingual-transfer \
-    --batch_size 16 \
+    --per_device_train_batch_size 16 \
     --model_name_or_path ernie-m-base \
     --save_steps 12272 \
-    --output_dir output
+    --num_train_epochs 5 \
+    --warmup_ratio 0.1 \
+    --remove_unused_columns False
 ```
```
Comment (wawltor) [inline model_zoo/ernie-m/README.md:162]: 这里的部署代码相对比较缺失比较多，
需要补充一下predict和server的代码

File: tests/fixtures/tests_samples/xnli/XNLI-1.0/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+This folder contains the "Cross-lingual Natural Language Inference" (XNLI) data.
+XNLI consists of dev and test sets for the MNLI task, in 15 languages:
```
Comment (wawltor) [inline tests/fixtures/tests_samples/xnli/XNLI-1.0/README.md]: 如果是从单测角度来看的，不需要这么多多语言文件，只需要1~2个即可

File: tests/fixtures/tests_samples/xnli/XNLI-1.0/xnli.dev.jsonl [hunk 3]
```diff
@@ -0,0 +1,300 @@
+{"annotator_labels": ["neutral", "contradiction", "neutral", "neutral", "neutral"], "genre": "facetoface", "gold_label": "neutral", "language": "ar", "match": "True", "pairID": "1", "promptID": "1", "sentence1": "\u0648\u0642\u0627\u0644\u060c \u0645\u0627\u0645\u0627\u060c \u0644\u0642\u062f \u0639\u062f\u062a \u0644\u0644\u0645\u0646\u0632\u0644.", "sentence1_tokenized": "\u0648\u0642\u0627\u0644 \u060c \u0645\u0627\u0645\u0627 \u060c \u0644\u0642\u062f \u0639\u062f\u062a \u0644\u0644\u0645\u0646\u0632\u0644 .", "sentence2": "\u0627\u062a\u0635\u0644 \u0628\u0623\u0645\u0647 \u062d\u0627\u0644\u0645\u0627 \u0623\u0648\u0635\u0644\u062a\u0647 \u062d\u0627\u0641\u0644\u0629 \u0627\u0644\u0645\u062f\u0631\u0633\u064a\u0629.", "sentence2_tokenized": "\u0627\u062a\u0635\u0644 \u0628\u0623\u0645\u0647 \u062d\u0627\u0644\u0645\u0627 \u0623\u0648\u0635\u0644\u062a\u0647 \u062d\u0627\u0641\u0644\u0629 \u0627\u0644\u0645\u062f\u0631\u0633\u064a\u0629 ."}
```
Comment (wawltor) [inline tests/fixtures/tests_samples/xnli/XNLI-1.0/xnli.dev.jsonl]: 需要严格限制一下单测文件的行数，10~20行即可，因为单测文件过大的话会导致PaddleNLP 整体的包过大

File: model_zoo/ernie-m/run_classifier.py [hunk 4]
```diff
@@ -189,58 +109,55 @@ def __len__(self):
...
-    train_data_loader = DataLoader(
-        dataset=train_ds, batch_sampler=train_batch_sampler, collate_fn=batchify_fn, num_workers=0, return_list=True
-    )
+    data_collator = DataCollatorWithPadding(tokenizer)
 
-    num_classes = 3
-    model = AutoModelForSequenceClassification.from_pretrained(
-        args.model_name_or_path, num_classes=num_classes, dropout=args.dropout
+    num_labels = 3
+    model = ErnieMForSequenceClassification.from_pretrained(
```
Comment (wawltor) [inline model_zoo/ernie-m/run_classifier.py]: 这里 AutoModelForSequenceClassification 跑不通？ 我记得是config导入有问题，这块需要测试一下

File: model_zoo/ernie-m/run_classifier.py [hunk 5]
```diff
@@ -250,94 +167,70 @@ def do_train(args):
...
+            test_ds = load_dataset("xnli", language, split="test")
+            test_ds = test_ds.map(trans_func, batched=True, remove_columns=remove_columns, load_from_cache_file=True)
+            metrics = trainer.evaluate(eval_dataset=test_ds)
+            metrics = {k + f"_{language}": v for k, v in metrics.items()}
+            combined.update({f"eval_accuracy_{language}": metrics.get(f"eval_accuracy_{language}", 0.0)})
+            trainer.log_metrics("eval", metrics)
+
+        combined.update({"eval_accuracy_average": np.mean(list(combined.values()))})
+        trainer.log_metrics("eval", combined)
+        trainer.save_metrics("eval", combined)
```
Comment (wawltor) [inline model_zoo/ernie-m/run_classifier.py:296]: 这里需要增加一下静态图导出的逻辑

Comment (Yam0214) [reply model_zoo/ernie-m/run_classifier.py:296]: <del>这里是说加上一个`trainer.save_state()`吗？</del>evaluation 也需要保存？
我再看看

Comment (Yam0214) [reply model_zoo/ernie-m/run_classifier.py:296]: 1. 已补充predictor和server的文件
2. 将`ErnieMForSequenceClassification`替换回`AutoModelForSequenceClassification`
3. 已补充静态图导出的逻辑
4. 小数据集控制在3行样本

File: scripts/regression/ci_case.sh [hunk 6]
```diff
@@ -979,16 +979,16 @@ if [ ! -f 'test.py' ];then
...
         --save_steps 2 \
         --max_steps 2 \
         --output_dir output \
         --logging_steps 1  >${log_path}/ernie-m >>${log_path}/ernie-m 2>&1
         print_info $? ernie-m
 else 
-    python -m pytest ${nlp_dir}/model_zoo/ernie-layout/ >${log_path}/ernie-layout >>${log_path}/ernie-layout 2>&1
-    print_info $? ernie-layout
+    python -m pytest ${nlp_dir}/model_zoo/ernie-m/ >${log_path}/ernie-m >>${log_path}/ernie-m 2>&1
+    print_info $? ernie-m
```
Comment (wj-Mcat) [inline scripts/regression/ci_case.sh]: 我建议上面通过脚本式和pytest两段脚本都要执行：

1. 保证老测试依旧 work、pytest 测试后面顺利迁移。
2. 等旧测试执行一段时间之后，再将其删除，从而保证其正确性。
3. pytest 中使用的是小批量模型和数据集，在 gpu 上面执行耗时少，为了同时保证新旧测试都能够正常运行，两段脚本都建议添加。

你在这里添加一个 todo，描述下通过命令行执行的方式都将要被删除，统一使用 pytest 来执行测试。

BTW：
在当前这个代码中，这段代码是永远不会执行的：在 `model_zoo/ernie_m/` 目录下是没有`test.py`文件的。

File: model_zoo/ernie-m/run_classifier.py [hunk 7]
```diff
@@ -250,94 +231,105 @@ def do_train(args):
...
+    # Define the metrics of tasks.
+    def compute_metrics(p):
+        # Define the metrics of tasks.
+        preds = p.predictions[0] if isinstance(p.predictions, tuple) else p.predictions
+
+        preds = paddle.to_tensor(preds)
+        label = paddle.to_tensor(p.label_ids)
+
+        metric = Accuracy()
+        metric.reset()
```
Comment (wj-Mcat) [inline model_zoo/ernie-m/run_classifier.py]: 因为你每次都是新建一个 metric 对象，所以这里没必要reset。

如果你把这个 metric 放在`compute_metrics`外面的话，只需要在函数的最后 reset 一下就行了。

File: model_zoo/ernie-m/README.md [hunk 8]
```diff
@@ -17,67 +25,145 @@
...
+- `do_train` 是否进行训练任务。
+- `do_eval` 是否进行评估任务。
+- `do_predict` 是否进行评测任务。
+- `do_export` 是否导出模型。
 - `output_dir` 表示模型保存路径。
+- `export_model_dir` 模型的导出路径。
+- `per_device_train_batch_size` 表示训练时每次迭代**每张**卡上的样本数目。
+- `per_device_eval_batch_size` 表示验证时每次迭代**每张**卡上的样本数目。
 - `max_seq_length` 表示最大句子长度，超过该长度将被截断，不足该长度的将会进行 padding。
-- `memory_length` 表示当前的句子被截取作为下一个样本的特征的长度。
```
Comment (ZHUI) [inline model_zoo/ernie-m/README.md:38]: 不需要了吗？

File: model_zoo/ernie-m/deploy/predictor/inference.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        "--model_name_or_path",
+        default="ernie-m-base",
+        type=str,
+        help="The directory or name of model.",
+    )
+    parser.add_argument(
+        "--model_path",
+        type=str,
+        required=True,
+        help="The path prefix of inference model to be used.",
```
Comment (ZHUI) [inline model_zoo/ernie-m/deploy/predictor/inference.py:46]: 看起来两个参数可以合并。**model_name_or_path** 只是给tokenizer使用，一般我们tokenizer也会保存一份。

不强制，跟其他例子保持一致即可。

File: model_zoo/ernie-m/run_classifier.py [hunk 10]
```diff
@@ -189,58 +142,86 @@ def __len__(self):
...
+def do_train():
+    training_args, model_args = PdArgumentParser([TrainingArguments, ModelArguments]).parse_args_into_dataclasses()
+    training_args: TrainingArguments = training_args
+    model_args: ModelArguments = model_args
+
+    training_args.print_config(model_args, "Model")
+
+    paddle.set_device(training_args.device)
     if paddle.distributed.get_world_size() > 1:
         paddle.distributed.init_parallel_env()
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py:152]: 删除

    if paddle.distributed.get_world_size() > 1:
        paddle.distributed.init_parallel_env()

File: model_zoo/ernie-m/run_classifier.py [hunk 11]
```diff
@@ -189,58 +142,86 @@ def __len__(self):
...
-    num_classes = 3
+    num_labels = 3
     model = AutoModelForSequenceClassification.from_pretrained(
-        args.model_name_or_path, num_classes=num_classes, dropout=args.dropout
+        model_args.model_name_or_path, num_labels=num_labels, classifier_dropout=model_args.classifier_dropout
     )
-    n_layers = model.ernie_m.config["num_hidden_layers"]
+    n_layers = model.config.num_hidden_layers
     if paddle.distributed.get_world_size() > 1:
         model = paddle.DataParallel(model)
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py]: 删除

File: model_zoo/ernie-m/run_classifier.py [hunk 12]
```diff
@@ -189,58 +142,86 @@ def __len__(self):
...
+        )
     else:
-        num_training_steps = len(train_data_loader) * args.num_train_epochs
-        num_train_epochs = args.num_train_epochs
-
-    warmup = args.warmup_steps if args.warmup_steps > 0 else args.warmup_proportion
-
-    lr_scheduler = LinearDecayWithWarmup(args.learning_rate, num_training_steps, warmup)
+        num_training_steps = 10
+    lr_scheduler = LinearDecayWithWarmup(training_args.learning_rate, num_training_steps, warmup)
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py]: 删除吧

File: model_zoo/ernie-m/run_classifier.py [hunk 13]
```diff
@@ -250,94 +231,103 @@ def do_train(args):
     for n, p in model.named_parameters():
         name_dict[p.name] = n
 
-    simple_lr_setting = partial(layerwise_lr_decay, args.layerwise_decay, name_dict, n_layers)
+    simple_lr_setting = partial(layerwise_lr_decay, model_args.layerwise_decay, name_dict, n_layers)
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py:244]: 这里的话，不建议这么用。

https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/optimizer/AdamW_cn.html#daimashili

image

设置内部的 params 的  learning_rate。造一个 params list 然后再set一下即可。

trainer.set_optimizer_grouped_parameters(params_to_train)

File: model_zoo/ernie-m/run_classifier.py [hunk 14]
```diff
@@ -250,94 +231,103 @@ def do_train(args):
...
+
+        metric = Accuracy()
+        result = metric.compute(preds, label)
+        metric.update(result)
+        accu = metric.accumulate()
+        return {"accuracy": accu}
+
+    trainer = Trainer(
+        model=model,
+        criterion=criterion,
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py]: 不要 criterion 了，模型里面支持 label 输入

File: model_zoo/ernie-m/run_classifier.py [hunk 15]
```diff
@@ -250,94 +231,103 @@ def do_train(args):
...
+        args=training_args,
+        data_collator=data_collator,
+        train_dataset=train_ds if training_args.do_train else None,
+        eval_dataset=eval_ds if training_args.do_eval else None,
+        tokenizer=tokenizer,
+        compute_metrics=compute_metrics,
+        optimizers=[optimizer, lr_scheduler],
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py:265]:
```suggestion

```
删除

File: model_zoo/ernie-m/run_classifier.py [hunk 16]
```diff
@@ -250,94 +231,103 @@ def do_train(args):
...
+        trainer.save_model()
+        trainer.log_metrics("train", metrics)
+        trainer.save_metrics("train", metrics)
+        trainer.save_state()
+
+    # Evaluating
+    if training_args.do_eval:
+        combined = {}
+        for language in all_languages:
+            eval_ds = load_xnli_dataset(model_args, "xnli", language, split="validation")
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py:287]: 这个 可以 支持一下 eval_ds dict输入

File: model_zoo/ernie-m/README.md (hunk 8 above)
Comment (Yam0214) [reply model_zoo/ernie-m/README.md:38]: 是的，`memory_length`没有被使用。

File: model_zoo/ernie-m/deploy/predictor/inference.py (hunk 9 above)
Comment (Yam0214) [reply model_zoo/ernie-m/deploy/predictor/inference.py:46]: 这里的`model_path`指向导出的静态图模型的路径，和`model_name_or_path`不太一样。暂且保留吧。

File: model_zoo/ernie-m/run_classifier.py (hunk 16 above)
Comment (Yam0214) [reply model_zoo/ernie-m/run_classifier.py:287]: 我看了HF的逻辑，如果我们支持 `trianer.evaluate(eval_dataset: Union[Dict[str, Dataset], Dataset, None])`，会和HF的行为不对齐。
<del>如果计划支持这个功能，我另起一个PR实现下。</del>

HF里 `trainer.evaluate()`认为传入的`eval_dataset`要么是`datasets.Dataset`，要么是`torch.utils.data.IterableDataset`（见[evaluate](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2871)和[get_eval_dataloader](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L925)）。如果在实例化trainer时`trainer.eval_dataset`传入了一个dict，那么在训练调用`trainer.evaluate`之前先做类型判断，拆出字典中的dataset循环送进`trainer.evaluate`（见[_maybe_log_save_evaluate](https://github.com/huggingface/transformers/blob/c836f77266be9ace47bff472f63caf71c0d11333/src/transformers/trainer.py#L2172)）。

File: scripts/regression/ci_case.sh (hunk 6 above)
Comment (Yam0214) [reply scripts/regression/ci_case.sh]: * 已修改`ci_case.sh`
* 删去`metric.reset()`

File: model_zoo/ernie-m/run_classifier.py (hunk 16 above)
Comment (ZHUI) [reply model_zoo/ernie-m/run_classifier.py:287]: 对 这块 确实不是这个场景。应该是 边训练 边 eval的时候支持多个 数据输入。

> 如果计划支持这个功能，我另起一个PR实现下。

有计划的，你可以帮忙加一下吗？张斌那边有这个场景

File: model_zoo/ernie-m/deploy/predictor/inference.py (hunk 9 above)
Comment (ZHUI) [reply model_zoo/ernie-m/deploy/predictor/inference.py:46]: `model_name_or_path` 其实同时支持了 `model_name` or `path`

File: model_zoo/ernie-m/run_classifier.py [hunk 17]
```diff
@@ -262,16 +219,41 @@ def compute_metrics(p):
...
+
+            if "encoder.layers" in static_name:
+                idx = static_name.find("encoder.layers.")
+                layer = int(static_name[idx:].split(".")[2])
+                ratio = layerwise_decay ** (n_layers - layer)
+            elif "embedding" in static_name:
+                ratio = layerwise_decay ** (n_layers + 1)
+
+            params_list.append({"params": param, "learning_rate": param.optimize_attr["learning_rate"] * ratio})
+        return params_list
```
Comment (ZHUI) [inline model_zoo/ernie-m/run_classifier.py]: `weight_decay` 和 `learning_rate` 是独立设置的。不能 continue 跳过。

这里 `param.optimize_attr["learning_rate"] `应该不需要吧？

**Review by ZHUI — APPROVED** [review]
LGTM
